### PR TITLE
FS-860 create account store

### PR DESCRIPTION
**Pull request by SentientHousePlant**
# Summary 

This PR initialises the account store API with the flask library[ connexion. ](https://engineering.zalando.com/posts/2016/12/crafting-effective-microservices-in-python.html)

> Because connexion is much newer and up to date then restx (it uses OpenAPI3 spec) the swagger ui looks different. This is nothing to be concerned about.

The api.yml contains the full spec of this api. If you dislike an endpoint name or anything or that matter you only need to edit the api.yml - no python needs to be edited! _21st century stuff._ 

# Details

The dummy db mirrors redis. 

The docker file reduces the workers in Gunicorn to 1 so the memory is guaranteed to be persisted between api calls. When we add a real db we can increase this back to 4.

### CODE REVIEW

File: core/account.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        account_id (_type_, optional): An account_id given as a string.
+        Defaults to None.
+
+    Returns:
+        dict, int
+    """
+
+    # print(f"email : {email_address} + account_id {account_id}")
+
+    if email_address is None and account_id is None:
```
Comment (endamccormack) [inline core/account.py:30]: is this a new line between everything an agreed style for us? when the code doesnt logically clump together I find it harder to read.

File: tests/test_accounts.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        exists in the account store.
+
+        GIVEN The flask test client
+        WHEN an email is submitted
+        THEN we expect that the email exists in the account store.
+
+        """
+
+        status_code, response_data = self.post_email_and_return_data(
+            "ram@harry-styles-fanclub.org"
```
Comment (endamccormack) [inline tests/test_accounts.py:26]: 😂

File: tests/test_accounts.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+        response = self.client.get(req.url)
+
+        return response
+
+    def post_email_and_return_data(
+        self, email_address: str
+    ) -> Tuple[int, dict]:
+
+        params = {"email_address": email_address}
```
Comment (endamccormack) [inline tests/test_accounts.py:127]: personal preference is with space, do we have a team standard for this?

{ "email_address" : email_address }

File: core/account.py (hunk 1 above)
Comment (endamccormack) [reply core/account.py:30]: generally i'd clump variables, logic and return related. For smaller methods I'd clump everything. Logic can be separated into multiple logical clumps when it makes logical sense to do so

File: openapi/api.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            type: string
+        - name: account_id
+          in: query
+          description: "The account id being used for the account id look up, if both account_id and email_address is given then the look up will use the account_id instead."
+          required: false
+          schema:
+            type: string
+      responses:
+        200:
+          description: The requested account exists, and its json and sent back as a response.
```
Comment (endamccormack) [inline openapi/api.yml:28]: we respond with its json payload maybe?

File: openapi/api.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/account'
+        204:
+          description: "The requested account doesn't exist."
+    post:
+      summary: "Post a new email - creating a json in the account store with a unique account id."
+      description: "Given an email as a query string an account_model json is created in the data store."
```
Comment (endamccormack) [inline openapi/api.yml:37]: confusing to read

File: openapi/api.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      parameters:
+      - name: email_address
+        in: query
+        description: "The email being used for the account creation."
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
+          description: "If the resource as successfully created then the created resource is sent back."
```
Comment (endamccormack) [inline openapi/api.yml:48]: returned, not sent back

File: openapi/api.yml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        200:
+          description: The requested account exists, and its json and sent back as a response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/account'
+        204:
+          description: "The requested account doesn't exist."
+    post:
+      summary: "Post a new email - creating a json in the account store with a unique account id."
```
Comment (endamccormack) [inline openapi/api.yml:36]: probably should not refer "json in account store", its a level of detail that I dont think helps someone reading the docs. They dont care how you store it, just that they can put data in and get data out

File: openapi/api.yml [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+info:
+  description: Account store API for DLUHC Funding Service Design
+  version: "0.1.0"
+  title: Funding Service Design - Account store
+paths:
+  /account:
+    get:
+      summary: "Return the account data given the account email address or account id."
+      description: "Given a GET request with the account id or email address given as a query string we return a reponse containing the data associated with the account."
```
Comment (endamccormack) [inline openapi/api.yml:11]: Confusing, please consider rewriting this for clarity

File: openapi/api.yml [hunk 9]
```diff
@@ -0,0 +1,82 @@
+openapi: "3.0.0"
+
+info:
+  description: Account store API for DLUHC Funding Service Design
+  version: "0.1.0"
+  title: Funding Service Design - Account store
+paths:
+  /account:
+    get:
+      summary: "Return the account data given the account email address or account id."
```
Comment (endamccormack) [inline openapi/api.yml:10]: for not given

Comment (endamccormack) [reply openapi/api.yml:10]: an account email, not the

**Review by endamccormack — CHANGES_REQUESTED** [review]

File: core/account.py (hunk 1 above)
Comment (SentientHousePlant) [reply core/account.py:28]: Typically I do clump logical parts together - my fault for turning off line numbers in my IDE. Now fixed :)

File: tests/test_accounts.py (hunk 3 above)
Comment (SentientHousePlant) [reply tests/test_accounts.py:108]: I agree with your standard - I'll have a chat with the team and see if there's a way to add this to black (or not depending on its popularity)

File: openapi/api.yml (hunk 7 above)
Comment (SentientHousePlant) [reply openapi/api.yml]: Agreed - I like low cognitive load docs.

File: tests/test_accounts.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        assert status_code == 200
+        assert "account_id" in response_data.keys()
+        assert "applications" in response_data.keys()
+        assert response_data["email_address"] == "test@delete_me.org"
+
+    def test_double_post_returns_409(self):
+
+        params = {"email_address": "test2@delete_me.org"}
+
+        req = PreparedRequest()
```
Comment (harryyo) [inline tests/test_accounts.py:38]: Nice - will use this

File: tests/test_accounts.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        req.prepare_url(url, params)
+
+        response1 = self.client.post(req.url)
+        response2 = self.client.post(req.url)
+
+        assert response1.status_code == 200
+        assert response2.status_code == 409
+
+    def test_get_methods_work(self):
+
```
Comment (harryyo) [inline tests/test_accounts.py:50]: Maybe some comments here could be handy - perhaps a little hard to follow what is being tested exactly (or why they should be returning what they are). Alternatively perhaps slightly more descriptive naming of returned values (rather than just email...account_id)?

File: tests/test_accounts.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    def get_to_non_existing_resource_returns_204(self):
+
+        response = self.get_and_return_data(
+            email_address="dfgdfjg@sdjlkjsf.org"
+        )
+
+        assert response.status_code == 204
+
+    def get_and_return_data(self, email_address=None, account_id=None):
+
```
Comment (harryyo) [inline tests/test_accounts.py:85]: Again a docstring of this helper function could be handy

Comment (harryyo) [reply tests/test_accounts.py:85]: Or they could go in a separate helpers.py file maybe?

File: tests/test_accounts.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        assert email_response_data == account_response_data
+        assert email_dict["account_id"] == account_id
+        assert account_dict["email_address"] == email
+        assert (
+            account_dict["email_address"]
+            == email_dict["email_address"]
+            == "test3@delete_me.org"
+        )
+
+    def get_to_non_existing_resource_returns_204(self):
```
Comment (harryyo) [inline tests/test_accounts.py:76]: This could follow the test_ naming convention of methods in this class (or remove test from the others)

File: utils/definitions.py [hunk 14]
```diff
@@ -0,0 +1,9 @@
+"""
```
Comment (harryyo) [inline utils/definitions.py:1]: Tricky to know where the project root definition function should be, though could be potentially confusing if the project root is found from its relation to a file in the utils directory (i.e. if utils is moved then the project root will be moved). Perhaps the project_root could be just defined where it is defined in the core/app.py?

**Comment by SentientHousePlant** [comment]
- [ ] Add todos to passed state.

**Review by harryyo — APPROVED** [review]

File: core/app.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+    connexion_app = connexion.FlaskApp(
+        __name__,
+        specification_dir=project_root_path + "/openapi/",
+        options=options,
+    )
+
+    flask_app = connexion_app.app
+
+    connexion_app.add_api(project_root_path + "/openapi/api.yml")
```
Comment (ramsharma-prog) [inline core/app.py:23]: I would use this 💯

**Review by ramsharma-prog — APPROVED** [review]

File: core/account.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        email_address (str, optional): An email address given as a string.
+        Defaults to None.
+        account_id (_type_, optional): An account_id given as a string.
+        Defaults to None.
+
+    Returns:
+        dict, int
+    """
+
+    if account_id:
```
Comment (endamccormack) [inline core/account.py:28]: clump

**Review by endamccormack — APPROVED** [review]
